cl/sentinel/gossip.go

```diff
@@ -285,30 +289,55 @@ type GossipSubscription struct {
 	host         peer.ID
 	ch           chan *GossipMessage
 	ctx          context.Context
+	expiration   atomic.Value // Unix nano for how much we should listen to this topic
+	subscribed   atomic.Bool
 
 	topic *pubsub.Topic
 	sub   *pubsub.Subscription
 
 	cf context.CancelFunc
 	rf pubsub.RelayCancelFunc
 
-	setup  sync.Once
 	stopCh chan struct{}
 }
 
-func (sub *GossipSubscription) Listen() (err error) {
-	sub.setup.Do(func() {
-		sub.stopCh = make(chan struct{}, 3)
-		sub.sub, err = sub.topic.Subscribe()
-		if err != nil {
-			err = fmt.Errorf("failed to begin topic %s subscription, err=%w", sub.topic.String(), err)
-			return
+func (sub *GossipSubscription) Listen() {
+	go func() {
```

Comment:
I feel check it every 100ms is a bit cpu wasting. can you check if following rewrite version is the same as your idea?
```golang
go func() {
  sub := retrySubscribeUntilSuccess()
  timer := time.After(expirationTime.Sub(time.Now()))
  defer timer.Stop()
  for {
    select{
    case <-s.stopCh:
       return
    case <-ctx.Done():
       return
    case <-timer:
        sub.Close()
        return
    }
  }
}

```